File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/Java/apitest/apitest.java
/*
 * $Id:$
 *
 * apitest.java
 * Copyright (c) 2009 NetApp, Inc. All rights reserved.
 * Specifications subject to change without notice.
 *
 * Sample for using the netapp.manage.* classes to 
 * any ONTAPI API
 */
import netapp.manage.*;
import java.io.*;

public class apitest {

    public static void usage() {
        System.out.println("\nUsage:\n\tapitest (options) <host> <user> "
                + "<password> <ONTAPI-name> [<paramname> <arg> ...]\n");
        System.out.println("Options:\n");
        System.out
                .println("\t-i\tAPI specified as XML input, on the command line\n");
        System.out
                .println("\t-I\tAPI specified as XML input, on standard input\n");
        System.out
                .println("\t-t {type}\tServer type(type = filer, dfm, ocum, agent)\n");
        System.out
                .println("\t-v {vfiler name | vserver name}\tFor vfiler-tunneling or vserver-tunneling \n");
        System.out.println("\t-r\tUse RPC transport (Windows)\n");
        System.out.println("\t-n\tUse HTTP\n");
        System.out.println("\t-p\t{port}\n");
        System.out.println("\t-x\tShow the XML input and output\n");
        System.out.println("\t-X\tShow the raw XML input and output\n");
        System.out.println("\t-c\tConnection timeout\n");
        System.out.println("\t-h\tUse Host equiv authentication mechanism\n");
        System.out.println("\t-K\t{keystore-file}       Client keystore file to use\n");
        System.out.println("\t-P\t{keystore-passwd}     Password to access the keystore file\n");
        System.out.println("\t-E\t{keystore-key-passwd} Password to access the private key in the keystore file\n");
        System.out.println("\t-Y\t{keystore-file-type}  Type of the keystore file (JKS/PKCS12). Default is JKS\n");
        System.out.println("\t-T\t{truststore-file}     Truststore file to use during server certificate verification\n");
        System.out.println("\t-S\tEnable server certificate verification\n");
        System.out.println("\t-H\tEnable Hostname verification\n\n");
        System.out.println("Note:\n");
        System.out.println("\tUse server type 'dfm' (-t dfm) for OnCommand Unified Manager server version 5.2 or earlier.");
        System.out.println("\tUse server type 'ocum' (-t ocum) for OnCommand Unified Manager server version 6.0 or later.\n\n");
        System.out.println("\tBy default username and password shall be used for client authentication.");
        System.out.println("\tSpecify -K option for using Certificate Based Authentication (CBA).");
        System.out.println("\tServer certificate and Hostname verification is disabled by default for CBA.");  
        System.out.println("\tDo not provide username and password for -h, -r or CBA\n");
        System.out.println("Examples:\n");
        System.out.println("\tapitest sweetpea root tryme system-get-version");
        System.out.println("\tapitest amana root meat quota-report volume vol0");
        System.out.println("\tapitest -t dfm -K my_keystore.jks -P keypasswd amana dfm-about");
        System.out.println("\tapitest -t dfm -K my_keystore.jks -P keypasswd -S -T my_truststore.jks amana dfm-about\n");
        System.exit(1);
    }

    public static void main(String[] args) {
        NaElement xi = null;
        NaElement xo;
        NaServer s;
        String server_type, user, password;
        int argcnt = 0, index, i = 0, ind = 1, showxml = 0;
        String read_xml = "";
        boolean dorpc, dovfiler, dofiler, dodfm, doocum, doagent, host_equiv;
        boolean dossl=true;
        boolean useTimeout = false;
        int inputxml = 0, use_port = -1;
        String Cur_Line, vfiler_name = null, platform_type;
        int timeout = 0;
        boolean useCBA = false;
        boolean verifyServerCert = false;
        boolean verifyHostname = false;
        String keyStoreFile, keyStorePasswd, keyPasswd, keyStoreType;
        String trustStoreFile;

        dofiler = dodfm = doocum = doagent = dorpc = dovfiler = host_equiv = false;
        keyStoreFile = keyStorePasswd = keyPasswd = trustStoreFile = null;
        keyStoreType = "JKS";

        if (args.length < 3)
            usage();

        index = args[0].indexOf('-');
        while (args[argcnt].startsWith("-")) {
            switch (args[argcnt].charAt(ind)) {
            case 'i':
                inputxml = 1;
                argcnt++;
                break;
            case 'I':
                inputxml = 2;
                argcnt++;
                break;
            case 't':
                server_type = args[argcnt + 1];
                if (server_type.equals("dfm"))
                    dodfm = true;
                else if (server_type.equals("ocum")) {
                    doocum = true;
                    dossl = true;
                }
                else if (server_type.equals("agent"))
                    doagent = true;
                else if (server_type.equals("filer"))
                    dofiler = true;
                argcnt = argcnt + 2;
                break;
            case 'v':
                vfiler_name = args[argcnt + 1];
                dovfiler = true;
                argcnt = argcnt + 2;
                break;
            case 'r':
                platform_type = (System.getProperty("os.name"));
                if (!platform_type.matches("(?i).*windows .*")) {
                    System.out.println("-r is valid only on windows\n");
                    System.exit(1);
                }
                dorpc = true;
                argcnt++;
                break;
            case 'n':
                dossl = false;
                argcnt++;
                break;
            case 'p':
                use_port = Integer.parseInt(args[argcnt + 1]);
                argcnt = argcnt + 2;
                break;
            case 'x':
                showxml = 1;
                argcnt++;
                break;
            case 'X':
                showxml = 2;
                argcnt++;
                break;
            case 'h':
                host_equiv = true;
                argcnt++;
                break;
            case 'c':
                useTimeout = true;
                timeout = Integer.parseInt(args[argcnt + 1]);
                argcnt = argcnt + 2;
                break;
            case 'K':
                useCBA = true;
                dossl = true;
                keyStoreFile = args[argcnt + 1];
                argcnt = argcnt + 2;
                break;
            case 'P':
                keyStorePasswd = args[argcnt + 1];
                argcnt = argcnt + 2;
                break;
            case 'E':
                keyPasswd = args[argcnt + 1];
                argcnt = argcnt + 2;
                break;
            case 'Y':
                keyStoreType = args[argcnt + 1];
                argcnt = argcnt + 2;
                break;
            case 'T':
                trustStoreFile = args[argcnt + 1];
                argcnt = argcnt + 2;
                break;
            case 'S':
                verifyServerCert = true;
                argcnt++;
                break;
            case 'H':
                verifyHostname = true;
                argcnt++;
                break;
            default:
                System.out.println("\nERROR::Invalid Option");
                usage();
            }
        }

        if ((args.length < 4) && (host_equiv == false && dorpc == false))
            usage();

        if ((dodfm || doocum) && dovfiler) {
            System.out.println("The -v option is not a valid option for OnCommand Unified Manager server.\n");
            System.exit(1);
        }
        if ((dodfm || doocum) && dorpc) {
            System.out.println("The -r option is not a vaid option for OnCommand Unified Manager server.\n");
            System.exit(1);
        }
        if (host_equiv && dorpc) {
            System.out.println("Invalid usage of authentication style. "
                    + "Do not use -r option and -h option together.\n");
            System.exit(1);
        }
	if (dorpc && useTimeout) {
		System.out.println("Connection timeout value cannot be set for RPC authentication style.\n");
		System.exit(1);
	}
        if (verifyHostname && !verifyServerCert) {
            System.out.println("Hostname verification cannot be enabled when server certificate verification is disabled.");
            System.exit(1); 
        }
        if (!(keyStoreType.equalsIgnoreCase("JKS") || keyStoreType.equalsIgnoreCase("PKCS12"))) {
            System.out.println("Invalid KeyStore type. Possible values are JKS and PKCS12.\n");
            System.exit(1);
        }
        /* Set the port to the host type specified */
        if (use_port == -1) {
            if (dodfm)
                use_port = (dossl) ? 8488 : 8088;
            else if (doocum)
                use_port = 443;
            else if (doagent)
                use_port = (dossl) ? 4093 : 4092;
            else
                use_port = (dossl) ? 443 : 80;
        }
        try {
            /* Vfiler tunnelling requires ONTAPI version 7.0 to work */
            /*
             * NaServer is called to connect to servers and invoke API's. The
             * argument passed should be: NaServer(hostname, major API version
             * number, minor API version number
             */
            if (vfiler_name != null)
                s = new NaServer(args[argcnt], 1, 7);
            else
                s = new NaServer(args[argcnt], 1, 0);

            /* This is needed for -X option */
            if (showxml == 2)
                s.setSnoop(-1);

            /*
             * Set the authentication style for subsequent ONTAPI
             * authentications.
             */
            if (dorpc)
                s.setStyle(NaServer.STYLE_RPC);
            else if (host_equiv) {
                s.setStyle(NaServer.STYLE_HOSTSEQUIV);
            } else if (useCBA)
                s.setStyle(NaServer.STYLE_CERTIFICATE);
            else
                s.setStyle(NaServer.STYLE_LOGIN_PASSWORD);

            /*
             * Set the login and password used for authenticating when an ONTAPI
             * API is invoked. When host_equiv is set do not set username and
             * password
             */
            if (!host_equiv && !dorpc && !useCBA) {
                user = args[++argcnt];
                if (args.length == argcnt + 1) {
                    System.out.println("\nInsufficient Arguments Passed");
                    usage();
                }
                password = args[++argcnt];
                s.setAdminUser(user, password);
            }

            /* Set the TCP port used for API invocations on the server. */
            s.setPort(use_port);

            if (useTimeout) {
                if (timeout > 0) {
                    s.setTimeout(timeout);
                } else {
                    System.out
                            .println("\nInvalid value for connection timeout. "
                                    + "Connection timeout value should be greater "
                                    + "than 0.\n");
                    System.exit(1);
                }
            }

            /* Set the Type of API Server */
            if (dodfm)
                s.setServerType(NaServer.SERVER_TYPE_DFM);
            else if (doocum)
                s.setServerType(NaServer.SERVER_TYPE_OCUM);
            else if (doagent)
                s.setServerType(NaServer.SERVER_TYPE_AGENT);
            else
                s.setServerType(NaServer.SERVER_TYPE_FILER);
            if (!dossl)
                s.setTransportType(NaServer.TRANSPORT_TYPE_HTTP);

            /*
             * Set the name of the vfiler on which the API commands need to be
             * invoked.
             */
            if (vfiler_name != null)
                s.setVfilerTunneling(vfiler_name);

            if (useCBA) {
                if (keyStorePasswd == null) {
                    System.out.println("Missing keystore password\n");
                    usage();
                }
                s.setKeyStore(keyStoreFile, keyStorePasswd, keyPasswd);
                if(keyStoreType.equalsIgnoreCase("JKS")) {
                    s.setKeyStoreType(NaServer.KEYSTORE_TYPE_JKS);
                } else {
                    s.setKeyStoreType(NaServer.KEYSTORE_TYPE_PKCS12);
                }
            }
            if (dossl || verifyServerCert) {
                if (verifyServerCert) {
                    s.enableServerCertVerification();
                } else {
                    s.disableServerCertVerification();
                }
                if (verifyServerCert) {
                    if (verifyHostname) {
                        s.enableHostnameVerification();
                    } else {
                        s.disableHostnameVerification();
                    }
                }
            }
            if (trustStoreFile != null) {
                    s.setTrustStore(trustStoreFile);
            }
            if (inputxml > 0) {
                if (inputxml == 2) {
                    if (args.length - 1 > argcnt) {
                        System.out.println("The -I option expects no API "
                                + "on the command-line, "
                                + "it expects standard input\n");
                        usage();
                    }
                    /* Read the input from standard in. */
                    InputStreamReader converter = new InputStreamReader(
                            System.in);
                    BufferedReader in = new BufferedReader(
                            new InputStreamReader(System.in));
                    while ((Cur_Line = in.readLine()) != null)
                        read_xml = read_xml + Cur_Line;
                } else {
                    if (args.length == argcnt + 1) {
                        System.out.println("API not specified");
                        usage();
                    }
                    for (int cnt = argcnt + 1; cnt < args.length; cnt++)
                        read_xml = read_xml + args[cnt];
                }
                args = null;
                args = read_xml.split(" |\n");
                read_xml = "";
                for (int cnt = 0; cnt < args.length; cnt++)
                    if (!(args[cnt].matches("\t") || args[cnt].matches(" ")))
                        read_xml = read_xml + args[cnt];
                args = null;
                args = read_xml.split("\t|\n");
                argcnt = -1;
            }
            /*
             * Invoke any ONTAPI API with arguments in (key,value) pair
             * args[0]=filer,args[1]=user,args[2]=passwd args[3]=Ontapi
             * API,args[4] onward arguments in (key,value) pair
             */
            if (inputxml > 0)
                xi = s.getXMLParseInput(args[++argcnt]);
            else
                xi = new NaElement(args[++argcnt]);
            try {
                if (args.length > argcnt + 1) {
                    for (int index2 = argcnt + 1; index2 < args.length; index2++) {
                        xi.addNewChild(args[index2], args[index2 + 1]);
                        index2++;
                    }
                }
            } catch (ArrayIndexOutOfBoundsException e) {
                throw new NaAPIFailedException(-1,
                        "Mismatch in arguments passed "
                                + "(in (key,value) Pair) to " + "Ontapi API");
            }
            if (showxml == 1) {
                System.out.println("INPUT:\n" + xi.toPrettyString(""));
            }
            /* Invoke a single ONTAPI API */
            xo = s.invokeElem(xi);
            if (showxml != 2)
                System.out.println("OUTPUT:\n" + xo.toPrettyString(""));
            s.close();
        } catch (NaAPIFailedException e) {
            System.err.println(e.toString());
            System.exit(1);
        } catch (Exception e) {
            System.err.println(e.toString());
            System.exit(1);
        }
    }
}

File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/DotNet/CSharp/apitest/apitest.cs
//============================================================//
//                                                            //
//                                                            //
// apitest.cs                                                 //
//                                                            //
// Exploratory application for ONTAPI APIs                    //
// It lets you call any ONTAPI API with named arguments       //
//    (essentially a command-line version of the zexplore     //
//     utility)                                               //
//                                                            //
// Copyright 2002-2010 NetApp, Inc. All rights                //
// reserved. Specifications subject to change without notice. // 
//                                                            //
// This SDK sample code is provided AS IS, with no support or //
// warranties of any kind, including but not limited to       //
// warranties of merchantability or fitness of any kind,      //
// expressed or implied.  This code is subject to the license //
// agreement that accompanies the SDK.                        //
//                                                            //
// tab size = 4                                               //
//                                                            //
//============================================================//

using System;
using System.Collections.Generic;
using System.Text;
using NetApp.Manage;

namespace Netapp.Utility
{
    class ApiTest
    {
        private static void Usage()
        {
            Console.WriteLine("\nUsage:\n\t apitest [options] <host> <user> "
                + "<password> <ONTAPI-name> [<param-name> <arg> ...]\n");
            Console.WriteLine("Options:\n");
            Console.WriteLine("\t -t {type} \t Server type(type = filer, dfm, ocum, agent)\n");
            Console.WriteLine("\t -v {vfiler name | vserver name}  For vfiler-tunneling or vserver-tunneling \n");
            Console.WriteLine("\t -r \t Use RPC transport\n");
            Console.WriteLine("\t -n \t Use HTTP\n");
            Console.WriteLine("\t -p {port} \t Override port to use\n");
            Console.WriteLine("\t -c {timeout} \t Connection timeout value in seconds\n");
            Console.WriteLine("\t -o {originator_id} \t Pass Originator Id\n");
            Console.WriteLine("\t -C {cert-file}    Location of the client certificate file\n");
            Console.WriteLine("\t -P {cert-passwd}  Password to access the certificate file\n");
            Console.WriteLine("\t -T {cert-store-name} Client certificate store name. The default is 'My' store\n");
            Console.WriteLine("\t -L {cert-store-loc}  Client certificate store location. The default is 'CurrentUser'\n");
            Console.WriteLine("\t -N {cert-name} Subject name of the client certificate in the certificate store\n");
            Console.WriteLine("\t -S \t Enable server certificate verification\n");
            Console.WriteLine("\t -H \t Enable hostname verification\n");
            Console.WriteLine("\t -i \t API specified as XML input, on the command line\n");
            Console.WriteLine("\t -I \t API specified as XML input, on standard input\n");
            Console.WriteLine("\t -x \t Show the XML input and output\n");
            Console.WriteLine("\t -X \t Show the raw XML input and output\n");
            Console.WriteLine("\t -h \t Use Host equiv authentication mechanism\n");
            Console.WriteLine("\nNote:");
            Console.WriteLine("\n        Use server type 'dfm' (-t dfm) for OnCommand Unified Manager server version 5.2 or earlier.");
            Console.WriteLine("\n        Use server type 'ocum' (-t ocum) for OnCommand Unified Manager server version 6.0 or later.\n");
            Console.WriteLine("\n        By default username and password shall be used for client authentication.");
            Console.WriteLine("        Specify either -C, -P or -S, -L, -N options for using Certificate Based Authentication (CBA).");
            Console.WriteLine("\n        Server certificate and Hostname verification is disabled by default for CBA.");
            Console.WriteLine("\n        Do not provide username and password for -h, -r or CBA options.");
            Console.WriteLine("\n        The username or UID of the user administering the storage systems can be passed");
            Console.WriteLine("        to ONTAP as originator-id for audit logging.\n");
            Console.WriteLine("Examples:\n");
            Console.WriteLine("\t apitest sweetpea root tryme "
                + "system-get-version");
            Console.WriteLine("\n\t apitest amana root meat quota-report volume vol0\n");
            Console.WriteLine("\n\t apitest -t dfm -C clientcert.pfx -P mypasswd amana dfm-about\n");
            Console.WriteLine("\n\t apitest -t dfm -T My -L CurrentUser -N ram amana dfm-about\n");
            Environment.Exit(1);
        }

        private static void GetXMLInput(int inputXML,ref string[] args,ref int index)
        {
            String readXML = "";
            
            if (inputXML == 1)
            {
                if (args.Length == index + 1)
                {
                    Console.WriteLine("API not specified");
                    Usage();
                }
                for (int cnt = index + 1; cnt < args.Length; cnt++)
                {
                    readXML = readXML + args[cnt];
                }
            }
            else
            {
                String curLine = "";
                Console.WriteLine("Enter the input XML:\n");
                while ((curLine = Console.ReadLine()) != null)
                {
                    readXML += curLine;
                }
            }

            args = readXML.Split(new Char[] { ' ', '\n','\t','\r' });

            readXML = "";
            for (int cnt = 0; cnt < args.Length; cnt++)
            {
                if (!(args[cnt].Contains("\t") || args[cnt].Contains(" ")))
                {
                    readXML = readXML + args[cnt];
                }
            }
            args = readXML.Split(new Char[] { '\t', '\n' });
            index = -1;
        }

        static void Main(string[] args)
        {
            NaServer s;
            NaElement xi = null;
            NaElement xo;
            NaServer.TRANSPORT_TYPE transportType = NaServer.TRANSPORT_TYPE.HTTPS;
            NaServer.SERVER_TYPE serverType = NaServer.SERVER_TYPE.FILER;
            NaServer.AUTH_STYLE authStyle = NaServer.AUTH_STYLE.LOGIN_PASSWORD;
            String vfiler = null;
            String originatorId = null;
            String type;
            int index = 0;
            int showXML = 0;
            int inputXML = 0;
            int port = -1;
            int timeOut = -1;
            Boolean status = true;
            Boolean useRPC = false;
            Boolean useHostsEquiv = false;
            Boolean useCBA = false;
            Boolean verifyServerCert = false;
            Boolean verifyHostname = false;
            String certFile = null;
            String certPasswd = null;
            NaServer.CERT_STORE_NAME storeName = NaServer.CERT_STORE_NAME.MY;
            NaServer.CERT_STORE_LOCATION storeLocation = NaServer.CERT_STORE_LOCATION.CURRENT_USER;
            String certName = null;

            if (args.Length < 3)
            {
                Usage();
            }
            index = 0;

            try
            {
                while (args[index].StartsWith("-"))
                {
                    switch (args[index][1])
                    {
                        case 't':
                            type = args[index + 1];
                            if (type.Equals("dfm"))
                            {
                                serverType = NaServer.SERVER_TYPE.DFM;
                            }
                            else if (type.Equals("ocum"))
                            {
                                serverType = NaServer.SERVER_TYPE.OCUM;
                                transportType = NaServer.TRANSPORT_TYPE.HTTPS;
                            }
                            else if (type.Equals("agent"))
                            {
                                serverType = NaServer.SERVER_TYPE.AGENT;
                            }
                            else if (type.Equals("filer"))
                            {
                                serverType = NaServer.SERVER_TYPE.FILER;
                            }
                            else
                            {
                                Console.WriteLine("\nERROR: Invalid Option for Server type.");
                                Usage();
                            }
                            index = index + 2;
                            break;
                        case 'v':
                            vfiler = args[index + 1];
                            index = index + 2;
                            break;
                        case 'r':
                            authStyle = NaServer.AUTH_STYLE.RPC;
                            useRPC = true;
                            index++;
                            break;
                        case 'n':
                            transportType = NaServer.TRANSPORT_TYPE.HTTP;
                            index++;
                            break;
                        case 'p':
                            status = Int32.TryParse(args[index + 1], out port);
                            if (status == false)
                            {
                                Console.WriteLine("\nERROR: Invalid port no.");
                                Usage();
                            }
                            index = index + 2;
                            break;
                        case 'i':
                            inputXML = 1;
                            index++;
                            break;
                        case 'I':
                            inputXML = 2;
                            index++;
                            break;
                        case 'x':
                            showXML = 1;
                            index++;
                            break;
                        case 'X':
                            showXML = 2;
                            index++;
                            break;
                        case 'h':
                            authStyle = NaServer.AUTH_STYLE.HOSTSEQUIV;
                            useHostsEquiv = true;
                            index++;
                            break;
                        case 'c':
                            status = Int32.TryParse(args[index + 1], out timeOut);
                            if (status == false || timeOut <= 0)
                            {
                                Console.WriteLine("\nERROR: Invalid timeout value.");
                                Usage();
                            }
                            index = index + 2;
                            break;
                        case 'o':
                            originatorId = args[index + 1];
                            index = index + 2;
                            break;
                        case 'C':
                            useCBA = true;
                            certFile = args[index + 1];
                            index =  index + 2;
                            break;
                        case 'P':
                            useCBA = true;
                            certPasswd = args[index + 1];
                            index = index + 2;
                           break;
                        case 'T':
                            String name = args[index + 1];
                            useCBA = true;
                            switch(name)
                            {
                                case "AuthRoot": 
                                    storeName = NaServer.CERT_STORE_NAME.AUTH_ROOT;
                                    break;
                                case "CertificateAuthority":
                                    storeName = NaServer.CERT_STORE_NAME.CERTIFICATE_AUTHORITY;
                                    break;
                                case "My":
                                    storeName = NaServer.CERT_STORE_NAME.MY;
                                    break;
                                case "Root":
                                    storeName = NaServer.CERT_STORE_NAME.ROOT;
                                    break;
                                case "TrustedPeople":
                                    storeName = NaServer.CERT_STORE_NAME.TRUSTED_PEOPLE;
                                    break;
                                default:
                                    Console.WriteLine("Invalid store name: " + name);
                                    Console.WriteLine("Valid store names are: ");
                                    Console.WriteLine("My - certificate store for personal certificates");
                                    Console.WriteLine("Root - certificate store for trusted root certificate authorities");
                                    Console.WriteLine("AuthRoot - certificate store for third-party certificate authorities");
                                    Console.WriteLine("CertificateAuthority - certificate store for intermediate certificate authorities");
                                    Console.WriteLine("TrustedPeople - certificate store for directly trusted people and resources\n");
                                    Usage();
                                    break;
                            }
                            index = index + 2;
                            break;
                        case 'L':
                            String location = args[index + 1];
                            useCBA = true;
                            switch (location)
                            {
                                case "CurrentUser":
                                    storeLocation = NaServer.CERT_STORE_LOCATION.CURRENT_USER;
                                    break;
                                case "LocalMachine":
                                    storeLocation = NaServer.CERT_STORE_LOCATION.LOCAL_MACHINE;
                                    break;
                                default:
                                    Console.WriteLine("Invalid store location: " + location);
                                    Console.WriteLine("Valid store locations are: ");
                                    Console.WriteLine("CurrentUser - certificate store used by the current user");
                                    Console.WriteLine("LocalMachine - certificate store assigned to the local machine");
                                    Usage();
                                    break;
                            }
                            index = index + 2;
                            break;
                        case 'N':
                            certName = args[index + 1];
                            useCBA = true;
                            index = index + 2;
                            break;
                        case 'S':
                            verifyServerCert = true;
                            index++;
                            break;
                        case 'H':
                            verifyHostname = true;
                            index++;
                            break;
                        default:
                            Console.WriteLine("\nERROR: Invalid Option.");
                            Usage();
                            break;
                    } //switch (args[index][1]) {
                } //while (args[index].StartsWith("-")){
            }
            catch (System.IndexOutOfRangeException)
            {
                Console.WriteLine("\nERROR: Invalid Arguments.");
                Usage();
            }
            if (authStyle == NaServer.AUTH_STYLE.LOGIN_PASSWORD &&
                            args.Length < 4)
            {
                Usage();
            }

            if (useHostsEquiv == true && useRPC == true)
            {
                Console.WriteLine("\nERROR: Invalid usage of authentication style. " +
                "Do not use -r option and -h option together.\n");
                System.Environment.Exit(1);
            }
			if (useRPC == true && timeOut != -1) {
				Console.WriteLine("\nERROR: Connection timeout value cannot be set for RPC authentication style.\n");
				Environment.Exit(1);
			}
            if (verifyHostname && !verifyServerCert)
            {
                Console.WriteLine("\nERROR: Hostname verification cannot be enabled when server certificate verification is disabled.\n");
                Environment.Exit(1);
            }
            if (useCBA)
            {
                transportType = NaServer.TRANSPORT_TYPE.HTTPS;
                authStyle = NaServer.AUTH_STYLE.CERTIFICATE;
            }
            else if (authStyle == NaServer.AUTH_STYLE.LOGIN_PASSWORD)
            {
                if (index == args.Length)
                {
                    Console.WriteLine("\nERROR: Host not specified.");
                    Usage();
                }
                if ((index + 1) == args.Length)
                {
                    Console.WriteLine("\nERROR: User not specified.");
                    Usage();
                }
                else if ((index + 2) == args.Length)
                {
                    Console.WriteLine("\nERROR: Password not specified.");
                    Usage();
                }
            }

            if (port == -1)
            {
                    switch (serverType)
                    {
                        default:
                        case NaServer.SERVER_TYPE.FILER:
                            port = (transportType == NaServer.TRANSPORT_TYPE.HTTP ? 80 : 443);
                            break;
                        case NaServer.SERVER_TYPE.DFM:
                            port = (transportType == NaServer.TRANSPORT_TYPE.HTTP ? 8088 : 8488);
                            break;
                        case NaServer.SERVER_TYPE.OCUM:
                            port = 443;
                            break;
                        case NaServer.SERVER_TYPE.AGENT:
                            port = (transportType == NaServer.TRANSPORT_TYPE.HTTP ? 4092 : 4093);
                            break;
                    }
            }

            try
            {
                //1. Create an instance of NaServer object.
                //NaServer is used to connect to servers and invoke API's.
                if (vfiler != null)
                {
                    // Vfiler tunnelling requires ONTAPI version 7.1 to work 
                    s = new NaServer(args[index], 1, 7);
                }
                else
                {
                    s = new NaServer(args[index], 1, 0);
                 }

                //2. Set the server type
                s.ServerType = serverType;

                //3. Set the transport type
                s.TransportType = transportType;

                //4. Set the authentication style for subsequent ONTAPI authentications.
                s.Style = authStyle;

                //5. Set the login and password used for authenticating when
                //an ONTAPI API is invoked.
                if (authStyle == NaServer.AUTH_STYLE.LOGIN_PASSWORD)
                {
                    s.SetAdminUser(args[++index], args[++index]);
                }

                //6. Set the port number
                s.Port = port;

                //7. Optional - set the vfiler name for vfiler tunneling
                if (vfiler != null)
                {
                    s.SetVfilerTunneling(vfiler);
                }

                // Check if originator_id is set
                if (originatorId != null)
                {
                    s.OriginatorId = originatorId;
                }

                //Set the request timeout.
                if (timeOut != -1)
                {
                    s.TimeOut = timeOut;
                }

                if (useCBA)
                {
                    if(certFile == null && certPasswd != null)
                    {
                        Console.WriteLine("\nERROR: Certificate file not specified.");
                        Usage();
                    }
                    if (certFile != null)
                    {
                        if (certPasswd != null)
                        {
                            s.SetClientCertificate(certFile, certPasswd);
                        }
                        else
                        {
                            s.SetClientCertificate(certFile);
                        }
                    }
                    else
                    {
                        s.SetClientCertificate(storeName, storeLocation, certName);
                    }
                }
                
                s.ServerCertificateVerification = verifyServerCert;
                if (verifyServerCert)
                {
                    s.HostnameVerification = verifyHostname;
                }
                s.Snoop = 1;
                /* Invoke any  ONTAPI API with arguments
                    * in (key,value) pair 
                    * args[0]=filer,args[1]=user,args[2]=passwd
                    * args[3]=Ontapi API,args[4] onward arguments
                    * in (key,value) pair
                */
                try
                {
                    if (inputXML == 0)
                    {
                        if ((index+1) == args.Length)
                        {
                            Console.WriteLine("\nERROR: API not specified.");
                            Usage();
                        }
                        //8. Create an instance of NaElement which contains the ONTAPI API request
                        xi = new NaElement(args[++index]);
                    }
                    //Only use this for debugging
                    else
                    {
                        GetXMLInput(inputXML, ref args, ref index);
                        xi = s.ParseXMLInput(args[++index]);
                    }
                        
                    if (args.Length > index + 1)
                    {
                        for (int index2 = index + 1;index2 < args.Length;index2++)
                        {
                            //9. Optional - add the child elements to the parent 
                            xi.AddNewChild(args[index2], args[index2 + 1]);
                            index2++;
                        }
                    }

                    // Only use this for debugging purpose
                    if (showXML > 0)
                    {
                        if (showXML == 1)
                        {
                            Console.WriteLine("INPUT:\n" + xi.ToPrettyString(""));
                        }
                        else
                        {
                            s.DebugStyle = NaServer.DEBUG_STYLE.PRINT_PARSE;
                        }
                    }
                }
                catch (System.IndexOutOfRangeException)
                {
                    throw new NaApiFailedException("Mismatch in arguments passed "
                                                       + "(in (key,value) Pair) to "
                                                       + "Ontapi API", -1);
                }
                
                // 10. Invoke a single ONTAPI API to the server.
                // The response is stored in xo.
                xo = s.InvokeElem(xi);

                // Only use this for debugging purpose
               if (showXML > 0)
                {
                    if (showXML == 2)
                    {
                        //Simply return because the NaServer will print the raw XML OUTPUT.
                        return;
                    }
                    Console.WriteLine("OUTPUT:");
                }                    

                //11. Print the ONTAPI API response that is returned by the server
                Console.WriteLine(xo.ToPrettyString(""));
            }
            catch (NaApiFailedException e)
            {
                Console.Error.WriteLine(e.Message);
            }
            catch(Exception e)
            {
                Console.Error.WriteLine(e.Message);
            }
        }
    }
}
